File: tests/test_ideals.py
"""Tests for ideal generators."""

import numpy as np
import pytest
import sympy as sp

from deepgroebner.ideals import *


R1, x, y, z = sp.ring('x,y,z', sp.FF(32003), 'grevlex')
R2, a, b, c, d = sp.ring('a,b,c,d', sp.QQ, 'lex')
R3, t, u, v = sp.ring('t,u,v', sp.FF(101), 'grlex')


def test_cyclic():
    F = cyclic(3)
    x, y, z = F[0].ring.gens
    assert F == [x + y + z, x*y + y*z + z*x, x*y*z - 1]


@pytest.mark.parametrize("ring, d, B", [
    (R1, 0, [R1.one]),
    (R2, 1, [a, b, c, d]),
    (R1, 2, [x**2, x*y, x*z, y**2, y*z, z**2]),
    (R3, 3, [t**3, t**2*u, t**2*v, t*u**2, t*u*v, t*v**2, u**3, u**2*v, u*v**2, v**3]),
])
def test_basis(ring, d, B):
    assert set(basis(ring, d)) == set(B)


@pytest.mark.parametrize("d, dist, constants, D", [
    (1, "weighted", False, np.array([0.0, 1.0])),
    (1, "weighted", True, np.array([0.5, 0.5])),
    (1, "uniform", True, np.array([0.25, 0.75])),
    (5, "weighted", False, np.array([0] + 5 * [0.2])),
    (5, "weighted", True, np.ones(6)/6),
    (5, "uniform", True, np.array([1, 3, 6, 10, 15, 21]) / 56.0),
    (3, "maximum", True, np.array([0.5, 0, 0, 0.5])),
    (3, "maximum", False, np.array([0, 0, 0, 1.0])),
    (3, 'uniform', False, np.array([0, 3, 6, 10]) / 19.0),
    (3, 'weighted', False, np.array([0, 1, 1, 1]) / 3.0),
])
def test_degree_distribution(d, dist, constants, D):
    degree_dist = degree_distribution(R1, d, dist=dist, constants=constants)
    assert np.array_equal(degree_dist, D)


def test_RandomBinomialIdealGenerator():
    ideal_gen = RandomBinomialIdealGenerator(3, 5, 5)
    ideal_gen.seed(123)
    x0, x1, x2 = ideal_gen.ring.gens
    F = [x0**2*x2 + 495*x2,
         x1*x2**4 + 5901*x0*x1*x2,
         x0**5 + 14384*x0**3*x2**2,
         x0**3*x1*x2 + 16417*x1**2*x2,
         x0**3*x1*x2 + 13109*x1**3*x2**2]    
    assert next(ideal_gen) == F


def test_RandomIdealGenerator():
    ideal_gen = RandomIdealGenerator(3, 5, 5, 0.5)
    ideal_gen.seed(123)
    x0, x1, x2 = ideal_gen.ring.gens
    F = [x0**3*x2**2 + 10689*x0**2*x1 + 12547*x1*x2**2,
         x1*x2**4 + 15388*x1**2*x2 + 22355*x1*x2**2,
         x0*x1*x2**2 + 4665*x0**3 + 15800*x0**2*x1,
         x0**3*x2**2 + 8782*x1**2*x2**3 + 15890*x1*x2**2,
         x1**2*x2 + 30687*x1**2]
    assert next(ideal_gen) == F

File: tests/test_ideals.cpp
#include <vector>

#include "gtest/gtest.h"

#include "ideals.h"
#include "polynomials.h"


TEST(cyclic, example) {
  Polynomial x = {{1, {1,0,0}}};
  Polynomial y = {{1, {0,1,0}}};
  Polynomial z = {{1, {0,0,1}}};
  Polynomial one = {{1, {0,0,0}}};
  std::vector<Polynomial> F = {x + y + z, x*y + y*z + z*x, x*y*z - one};
  EXPECT_EQ(cyclic(3), F);
}


TEST(basis, example1) {
  std::vector<Monomial> B = {{}};
  EXPECT_EQ(basis(3, 0), B);
}


TEST(basis, example2) {
  std::vector<Monomial> B = {{1,0,0,0}, {0,1,0,0}, {0,0,1,0}, {0,0,0,1}};
  EXPECT_EQ(basis(4, 1), B);
}


TEST(basis, example3) {
  std::vector<Monomial> B = {{2,0,0}, {1,1,0}, {1,0,1}, {0,2,0}, {0,1,1}, {0,0,2}};
  EXPECT_EQ(basis(3, 2), B);
}


TEST(basis, example4) {
  std::vector<Monomial> B = {{3,0,0}, {2,1,0}, {2,0,1}, {1,2,0}, {1,1,1}, {1,0,2}, {0,3,0}, {0,2,1}, {0,1,2}, {0,0,3}};
  EXPECT_EQ(basis(3, 3), B);
}


TEST(binomial, example) {
  EXPECT_EQ(binomial(5, 0), 1);
  EXPECT_EQ(binomial(5, 1), 5);
  EXPECT_EQ(binomial(5, 2), 10);
  EXPECT_EQ(binomial(5, 3), 10);
  EXPECT_EQ(binomial(5, 4), 5);
  EXPECT_EQ(binomial(5, 5), 1);
}


TEST(degree_distribution, example1) {
  auto degree_dist = degree_distribution(3, 1, DistributionType::Weighted, false);
  std::vector<double> D = {0.0, 1.0};
  EXPECT_EQ(degree_dist.probabilities(), D);
}


TEST(degree_distribution, example2) {
  auto degree_dist = degree_distribution(3, 1, DistributionType::Weighted, true);
  std::vector<double> D = {0.5, 0.5};
  EXPECT_EQ(degree_dist.probabilities(), D);
}


TEST(degree_distribution, example3) {
  auto degree_dist = degree_distribution(3, 1, DistributionType::Uniform, true);
  std::vector<double> D = {0.25, 0.75};
  EXPECT_EQ(degree_dist.probabilities(), D);
}


TEST(degree_distribution, example4) {
  auto degree_dist = degree_distribution(3, 5, DistributionType::Weighted, false);
  std::vector<double> D = {0.0, 0.2, 0.2, 0.2, 0.2, 0.2};
  EXPECT_EQ(degree_dist.probabilities(), D);
}


TEST(degree_distribution, example5) {
  auto degree_dist = degree_distribution(3, 5, DistributionType::Weighted, true);
  std::vector<double> D = {1.0/6, 1.0/6, 1.0/6, 1.0/6, 1.0/6, 1.0/6};
  EXPECT_EQ(degree_dist.probabilities(), D);
}


TEST(degree_distribution, example6) {
  auto degree_dist = degree_distribution(3, 5, DistributionType::Uniform, true);
  std::vector<double> D = {1.0/56, 3.0/56, 6.0/56, 10.0/56, 15.0/56, 21.0/56};
  EXPECT_EQ(degree_dist.probabilities(), D);
}


TEST(degree_distribution, example7) {
  auto degree_dist = degree_distribution(3, 3, DistributionType::Maximum, true);
  std::vector<double> D = {0.5, 0.0, 0.0, 0.5};
  EXPECT_EQ(degree_dist.probabilities(), D);
}


TEST(degree_distribution, example8) {
  auto degree_dist = degree_distribution(3, 3, DistributionType::Maximum, false);
  std::vector<double> D = {0.0, 0.0, 0.0, 1.0};
  EXPECT_EQ(degree_dist.probabilities(), D);
}


TEST(degree_distribution, example9) {
  auto degree_dist = degree_distribution(3, 3, DistributionType::Uniform, false);
  std::vector<double> D = {0.0, 3.0/19, 6.0/19, 10.0/19};
  EXPECT_EQ(degree_dist.probabilities(), D);
}


TEST(degree_distribution, example10) {
  auto degree_dist = degree_distribution(3, 3, DistributionType::Weighted, false);
  std::vector<double> D = {0.0, 1.0/3, 1.0/3, 1.0/3};
  EXPECT_EQ(degree_dist.probabilities(), D);
}


TEST(RandomBinomialIdealGenerator, example1) {
  RandomBinomialIdealGenerator ideal_gen{3, 5, 5};
  ideal_gen.seed(123);
  std::vector<Polynomial> F = {{{1, {0,1,4}}, {31, {0,3,1}}},
			       {{1, {3,1,1}}, {16013, {3,0,2}}},
			       {{1, {2,2,0}}, {18427, {1,0,1}}},
			       {{1, {2,0,3}}, {15139, {2,1,1}}},
			       {{1, {0,3,2}}, {5374, {1,0,2}}}};
  EXPECT_EQ(ideal_gen.next(), F);
}


TEST(RandomIdealGenerator, example1) {
  RandomIdealGenerator ideal_gen{3, 5, 5, 0.5};
  ideal_gen.seed(123);
  std::vector<Polynomial> F = {{{1, {0,1,3}}, {22264, {0,0,4}}},
			       {{1, {1,1,1}}, {1541, {0,0,2}}},
			       {{1, {2,2,1}}, {15981, {0,2,1}}, {7023, {0,0,1}}},
			       {{1, {1,4,0}}, {10365, {0,5,0}}, {5289, {1,3,0}}, {13942, {1,1,0}}},
			       {{1, {3,1,0}}, {11636, {1,1,0}}}};
  EXPECT_EQ(ideal_gen.next(), F);
}
